virtwho/ssh.py

```diff
@@ -1,3 +1,152 @@
+            return self.pwd_transfer()
+        elif self.rsa:
+            return self.rsa_transfer()
+        else:
+            raise ConnectionError(self.err)
+
+    def pwd_connect(self):
+        """SSH command execution connection by password"""
+        ssh = paramiko.SSHClient()
+        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        ssh.connect(
+            self.host,
+            self.port,
+            self.user,
+            self.pwd,
+            timeout=self.timeout)
+        return ssh
+
+    def rsa_connect(self):
+        """SSH command execution connection by key file"""
+        pkey = paramiko.RSAKey.from_private_key_file(self.rsa)
+        ssh = paramiko.SSHClient()
+        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        ssh.connect(
+            self.host,
+            self.port,
+            self.user,
+            pkey=pkey,
+            timeout=self.timeout)
+        return ssh
+
+    def pwd_transfer(self):
+        """Sftp download/upload execution connection by password"""
+        transport = paramiko.Transport(self.host, self.port)
+        transport.connect(username=self.user, password=self.pwd)
+        sftp = paramiko.SFTPClient.from_transport(transport)
+        return sftp, transport
+
+    def rsa_transfer(self):
+        """Sftp download/upload execution connection by key file"""
+        pkey = paramiko.RSAKey.from_private_key_file(self.rsa)
+        transport = paramiko.Transport(self.host, self.port)
+        transport.connect(username=self.user, pkey=pkey)
+        sftp = paramiko.SFTPClient.from_transport(transport)
+        return sftp, transport
+
+    def runcmd(self, cmd):
+        """Executes SSH command on remote hostname.
+        :param str command: The command to run
+        """
+        ssh = self._connect()
+        logger.info(">>> {}".format(cmd))
+        stdin, stdout, stderr = ssh.exec_command(cmd)
+        code = stdout.channel.recv_exit_status()
+        stdout, stderr = stdout.read(), stderr.read()
+        ssh.close()
+        if not stderr:
+            logger.info("<<< stdout\n{}".format(stdout.decode()))
+            return code, stdout.decode()
+        else:
+            logger.info("<<< stderr\n{}".format(stderr.decode()))
+            return code, stderr.decode()
+
+    def get_file(self, remote_file, local_file):
+        """Download a remote file to the local machine.
+        """
+        sftp, conn = self._transfer()
+        sftp.get(remote_file, local_file)
+        conn.close()
+
+    def put_file(self, local_file, remote_file):
+        """Upload a local file to a remote machine
+        :param local_file: either a file path or a file-like object to be uploaded.
+        :param remote_file: a remote file path where the uploaded file will be placed.
+        """
+        sftp, conn = self._transfer()
+        sftp.put(local_file, remote_file)
+        conn.close()
+
+    def remove_file(self, remote_file_path):
```

Comment:
About the argument `remote_file_path`, is it a file or a dir?
If it's a file, we may keep the uniform name as `remote_file`,
If it's dir, we may keep the uniform name as 'remote_dir`.

Reply:
Oh, will change it, I even forgot added this function and didn't check the name.

Reply:
Changed it to remote_file, thanks.